ReduxSimple/ReduxStoreWithHistory.cs

```diff
@@ -0,0 +1,58 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace ReduxSimple
+{
+    public abstract class ReduxStoreWithHistory<TState> : ReduxStore<TState>
+        where TState : class, new()
+    {
+        private readonly Stack<TState> _previousStates = new Stack<TState>();
+        private readonly Stack<TState> _nextStates = new Stack<TState>();
+
+        public bool CanUndo => _previousStates.Count != 0;
+
+        public bool CanRedo => _nextStates.Count != 0;
+
+        protected ReduxStoreWithHistory(TState initialState = null) : base(initialState)
+        { }
+
+        public override void Dispatch(object action)
+        {
+            _nextStates.Clear();
+            _previousStates.Push(State);
+
+            base.Dispatch(action);
+        }
+
+        public bool Undo()
+        {
+            if (!CanUndo)
+            {
+                return false;
+            }
+
+            _nextStates.Push(State);
+
+            GoToState(_previousStates.Pop());
```

Comment:
We need to know what devs think will happen when using the `Undo` method. 

Should we trigger a new `OnNext` event for the State and the Action? Does it make it sense because to me it will like we continue the flow of actions forward and not going backward? Should we introduce new `Subject<Action>` per `Undo` call so devs will handle the change by themselves?

Reply:
I was thinking that triggering a new event for Action would be kind of awkward because we're not technically performing the action again when we undo. Your idea for having a separate `Subject<Action>` for `Undo` calls makes sense though.